pkg/metrics/metrics.go

```diff
@@ -0,0 +1,90 @@
+package metrics
+
+import (
+	"net/http"
+	"strings"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+)
+
+const (
+	// Default values which are to be used as labels
+	// for prometheus metrics.
+	ERROR      = "opsrc_error"
+	OPSRC_NAME = "opsrc_name"
+
+	// Path, Port and Host where custom metrics would be exposed.
+	METRICSPATH = "/metrics"
+	METRICSPORT = "8383"
+	METRICSHOST = "0.0.0.0"
+
+	// Error message to be returned when quay.io is down.
+	UNREACHABLE_ERROR = "Quay.io is unavailabe"
+
+	// Error message to be returned when host address is wrong.
+	NO_HOST_EXSISTS_ERROR = "No such host exsists. Check the quay endpoint"
+)
+
+// Metric variables that report error based on the operator-source
+// which fails to reconcile.
+var (
+	operatorSourceError = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "operator_source_failure_count",
+			Help: "Monotonic count of times operator-source could not download packages",
+		},
+		[]string{OPSRC_NAME, ERROR},
+	)
+	// List of custom metrics which are to be collected.
+	metricsList = []prometheus.Collector{
+		operatorSourceError,
+	}
+)
+
+// registerMetrics registers the metrics with prometheus.
+func registerMetrics() error {
+	for _, metric := range metricsList {
+		err := prometheus.Register(metric)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// RecordMetrics starts the server, records the custom metrics and exposes
+// the metrics at the specified endpoint.
+func RecordMetrics() {
+	// Register metrics for the operator with the prometheus.
+	registerMetrics()
+
+	// Start the server and expose the registered metrics.
+	http.Handle(METRICSPATH, promhttp.Handler())
+	port := ":" + METRICSPORT
+	go http.ListenAndServe(port, nil)
+}
+
+// counterForOpsrcError returns the number of times error was encountered while
+// contacting quay for operstor-sources.
+func counterForOpsrcError(opsrc, opsrcError string) prometheus.Counter {
+	return operatorSourceError.WithLabelValues(opsrc, opsrcError)
+}
+
+// updateMetrics takes the name of the operator source which fails to reconcile, and
+// increments the error count of the respective operator metric.
+func UpdateMetrics(opsrc, opsrcError string) {
```

Comment:
nit: to make it clear that that you should not pass any old value here in the future, it might make sense to pass `opsrcError` as an enum value instead of an arbitrary string.

Reply:
Actually, I am taking `opsrcError` based on the error message we get from `ListPackages` during Configuring phase. So it can be either of the three values (no_host, quay_down, or a custom error message) as of now. Did you mean that we pass the label values as constants ?

Reply:
I meant something along the lines of 

```go
type OpSrcError string

const (
        NO_HOST_ERROR OpSrcError = "no_host"
        UNREACHABLE_ERROR OpSrcError = "quay_unavailable"
)

func UpdateMetrics(opsrc, opSrcErr OpSrcError) {
  // ...
}
```

That way it's super clear to anyone calling `UpdateMetrics` that you shouldn't make up your own string and pass it in.

Reply:
Thanks @ecordell. Got it..